### DISCUSSION TITLE
Fundamentos do ASP.NET 6 - CRUD e EF - Duvida a respeito de organização de código

### DISCUSSION THREAD

**Discussion by lucastatagiba** [balta-io/forum#477] [Tira Dúvidas] [answered]
Olá bom dia.
Duvida:
Repositórios são a melhor forma para buscar dados de outra tabela se precisar num controller de outra tabela ? se sim, dai te pergunto, qual é o processo então ? os repositórios é q fazem as buscas no banco ou o controller q faz, ou os 2? pq não faz sentido ficar persistências nos 2(controllers e repositórios)... dai onde fica a regra de negocio e onde fica as persistências e buscas de dados no banco ?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @lucastatagiba, como vai?

Repositórios são a melhor forma para buscar dados de outra tabela se precisar num controller de outra tabela ? 
- Uma das formas sim, a melhor depende do que precisa!

se sim, dai te pergunto, qual é o processo então ? os repositórios é q fazem as buscas no banco ou o controller q faz, ou os 2?
- Os repositórios são responsáveis por persistir os dados de Domain Models, em um cenário com DDD por exemplo, sim seriam eles, mas a chamada a persistência talvez nem ficasse no controller.
- Neste caso, você poderia criar um objeto mais simples para isso, um DAO (Que funciona como um repositório, só que sem o foco em agregados).


 pq não faz sentido ficar persistências nos 2(controllers e repositórios)... 
- Realmente, não faz e não há necessidade. Pode manter só nos "repositórios"

dai onde fica a regra de negocio e onde fica as persistências e buscas de dados no banco ?
- Depende da arquitetura que está usando... se for MVC fica nos controladores mesmo
- Lembrando que o seu DbContext já funciona como um repositório.

**Reply by lucastatagiba** [reply]
Entendi, então quer dizer que pode haver um cenário onde trabalhamos somente com controller + DbContext sem ser necessário o uso do repositório em si ? Nesse caso é o próprio MVC, certo ? E outro cenário onde usamos o DDD que daí teríamos o repositório + um DAO…

**Reply by andrebaltieri** [reply]
Bom dia, @lucastatagiba, como vai?

"então quer dizer que pode haver um cenário onde trabalhamos somente com controller + DbContext sem ser necessário o uso do repositório em si?
- Sim, inclusive eu mostro isto no segundo e terceiro módulo do curso de Fundamentos do ASP.NET:
👉 https://balta.io/cursos/fundamentos-aspnet

"E outro cenário onde usamos o DDD que daí teríamos o repositório + um DAO…"
- Somente repositório, como mostro nos cursos da carreira avançada:
- https://balta.io/carreiras/advanced-dotnet-developer
